### PR TITLE
decompile unk_02008DEC.s

### PR DESCRIPTION
**Pull request by remicalixte**
No idea of what this is about but GF sure loves copy-pasting stuff.

### CODE REVIEW

File: arm9/src/unk_02008DEC.c [hunk 1]
```diff
@@ -0,0 +1,1122 @@
...
+    GF_ASSERT(param1);
+    GF_ASSERT(param1->unk04 > param2);
+    GF_ASSERT(param0->unk10 == param1->unk08);
+
+    struct UnkStruct_02008DEC_sub *st1c = FUN_020097D0(param0);
+    GF_ASSERT(st1c);
+
+    if (param1->unk0C == 0)
+    {
+        struct UnkStruct_02008DEC_3 *r6 = &((struct UnkStruct_02008DEC_3 *)param1->unk00)[param2];
```
Comment (red031000) [inline arm9/src/unk_02008DEC.c:77]: this cast feels slightly awkward

File: arm9/src/unk_02008DEC.c [hunk 2]
```diff
@@ -0,0 +1,1122 @@
...
+        GF_ASSERT(FUN_0200945C(param0, r6->unk00) == 1);
+
+        FUN_0200994C(
+            param0, st1c, r6->unk04, r6->unk00, r6->unk44, r6->unk48, param1->unk08, param3);
+    }
+    else
+    {
+        struct UnkStruct_02008DEC_4 *r6 = &((struct UnkStruct_02008DEC_4 *)param1->unk00)[param2];
```
Comment (red031000) [inline arm9/src/unk_02008DEC.c:85]: same with this cast

File: arm9/src/unk_02008DEC.c [hunk 3]
```diff
@@ -0,0 +1,1122 @@
...
+        ptr = AllocFromHeap(param2, param1->unk04 * sizeof(struct UnkStruct_02008DEC_4));
+    }
+    else
+    {
+        ptr = NULL;
+    }
+
+    param1->unk00 = ptr;
+    // use volatile to force a reload here
+    void *r0 = *((void *volatile *)param1);
```
Comment (red031000) [inline arm9/src/unk_02008DEC.c:574]: again this feels awkward, fakematch?

File: arm9/src/unk_02008DEC.c [hunk 4]
```diff
@@ -0,0 +1,1122 @@
...
+
+THUMB_FUNC BOOL FUN_0200973C(struct UnkStruct_02008DEC_2 *param0, s32 param1)
+{
+    GF_ASSERT(param0);
+    GF_ASSERT(param0->unk04 > param1);
+
+    BOOL r6;
+    if (param0->unk0C == 1)
+    {
+        r6 = ((struct UnkStruct_02008DEC_4 *)param0->unk00)[param1].unk08;
```
Comment (red031000) [inline arm9/src/unk_02008DEC.c:634]: are you sure about these casts?

**Review by red031000 — CHANGES_REQUESTED** [review]
a few things that seem a little bit... off

File: arm9/src/unk_02008DEC.c (hunk 3 above)
Comment (remicalixte) [reply arm9/src/unk_02008DEC.c:574]: indeed it is

**Comment by remicalixte** [comment]
About the weird casts, if you look at `FUN_02008E6C` there is an `if` where one branch does
```
        struct UnkStruct_02008DEC_3 *r6 = &((struct UnkStruct_02008DEC_3 *)param1->unk00)[param2];
```
and the other one 
```
        struct UnkStruct_02008DEC_4 *r6 = &((struct UnkStruct_02008DEC_4 *)param1->unk00)[param2];
```
The same struct field is accessed as a array to something. The array index gives us the size of the things in the array (in asm offset = index * size). This led me to create these two different structs definition because the sizes were clearly different: `sizeof(struct UnkStruct_02008DEC_3)` is 0x4C and `sizeof(struct UnkStruct_02008DEC_4)` is 0x18. If you know a better cleaner way to achieve the same result please tell me.

**Comment by PikalaxALT** [comment]
> If you know a better cleaner way to achieve the same result please tell me.

Google the `union` keyword in C.

**Comment by remicalixte** [comment]
I didn't think about using a `union` lol

**Review by red031000 — APPROVED** [review]

**Pull request merged**